evm/src/cpu/kernel/asm/fields/fp6.asm

```diff
@@ -0,0 +1,447 @@
+    // stack:                 x4, x5
+    %i9
+    // stack:                 y5, y4
+    PUSH $offset
+    %add_const(1)
+    %mstore_kernel_general
+    // stack:                     y4
+    PUSH $offset
+    %mstore_kernel_general
+    // stack:
+%endmacro
+
+// cost: 9; note this returns y, x for x + yi
+%macro i9
+    // stack:          a , b
+    DUP2
+    DUP2
+    // stack:  a , b,  a , b
+    PUSH 9
+    MULFP254
+    SUBFP254
+    // stack: 9a - b,  a , b
+    SWAP2 
+    // stack:  b , a, 9a - b
+    PUSH 9
+    MULFP254
+    ADDFP254
+    // stack: 9b + a, 9a - b 
+%endmacro
+
+// cost: 6
+%macro dup1_fp6
+    // stack:       f: 6
+    DUP6
+    DUP6
+    DUP6
+    DUP6
+    DUP6
+    DUP6
+    // stack: f: 6, g: 6
+%endmacro 
+
+// cost: 6
+%macro dup2_fp6
+    // stack:       f: 6, g: 6
+    DUP12
+    DUP12
+    DUP12
+    DUP12
+    DUP12
+    DUP12
+    // stack: g: 6, f: 6, g: 6
+%endmacro
+
+// cost: 16
+%macro swap_fp6
+    // stack: f0, f1, f2, f3, f4, f5, g0, g1, g2, g3, g4, g5
+    SWAP6
+    // stack: g0, f1, f2, f3, f4, f5, f0, g1, g2, g3, g4, g5
+    SWAP1
+    SWAP7
+    SWAP1
+    // stack: g0, g1, f2, f3, f4, f5, f0, f1, g2, g3, g4, g5
+    SWAP2
+    SWAP8
+    SWAP2
+    // stack: g0, g1, g2, f3, f4, f5, f0, f1, f2, g3, g4, g5
+    SWAP3
+    SWAP9
+    SWAP3    
+    // stack: g0, g1, g2, g3, f4, f5, f0, f1, f2, f3, g4, g5
+    SWAP4
+    SWAP10
+    SWAP4
+    // stack: g0, g1, g2, g3, g4, f5, f0, f1, f2, f3, f4, g5
+    SWAP5
+    SWAP11
+    SWAP5
+    // stack: g0, g1, g2, g3, g4, g5, f0, f1, f2, f3, f4, f5
+%endmacro
+
+// cost: 16
+%macro add_fp6
+    // stack: f0, f1, f2, f3, f4, f5, g0, g1, g2, g3, g4, g5
+    SWAP7
+    ADDFP254
+    SWAP6
+    // stack: f0,     f2, f3, f4, f5, g0, h1, g2, g3, g4, g5 
+    SWAP7
+    ADDFP254
+    SWAP6
+    // stack: f0,         f3, f4, f5, g0, h1, h2, g3, g4, g5 
+    SWAP7
+    ADDFP254
+    SWAP6
+    // stack: f0,             f4, f5, g0, h1, h2, h3, g4, g5
+    SWAP7
+    ADDFP254
+    SWAP6
+    // stack: f0,                 f5, g0, h1, h2, h3, h4, g5
+    SWAP7
+    ADDFP254
+    SWAP6
+    // stack: f0,                     g0, h1, h2, h3, h4, h5
+    ADDFP254
+    // stack:                         h0, h1, h2, h3, h4, h5
+%endmacro
+
+// *reversed argument subtraction* cost: 17
+%macro bus_fp6
+    // stack: f0, f1, f2, f3, f4, f5, g0, g1, g2, g3, g4, g5
+    SWAP7
+    SUBFP254
+    SWAP6
+    // stack: f0,     f2, f3, f4, f5, g0, h1, g2, g3, g4, g5 
+    SWAP7
+    SUBFP254
+    SWAP6
+    // stack: f0,         f3, f4, f5, g0, h1, h2, g3, g4, g5 
+    SWAP7
+    SUBFP254
+    SWAP6
+    // stack: f0,             f4, f5, g0, h1, h2, h3, g4, g5
+    SWAP7
+    SUBFP254
+    SWAP6
+    // stack: f0,                 f5, g0, h1, h2, h3, h4, g5
+    SWAP7
+    SUBFP254
+    SWAP6
+    // stack: f0,                     g0, h1, h2, h3, h4, h5
+    SWAP1
+    SUBFP254
+    // stack:                         h0, h1, h2, h3, h4, h5
+%endmacro
+
+// cost: 156
+%macro mul_fp6
```

Comment:
This is really big, so it should be a function, not a macro.

A macro gets copy-pasted a bunch of times, increasing the kernel size. That’s fine if it’s small, but if it’s big enough, then the overhead of a function call is relatively minimal.